Add features to GUI

Xpl0itR opened the pull request:
### TODO list for this PR
- [x] Change controller image depending on type of controller
- [x] Game list should scan subdirs for games
- [x] Keys file exists checker
- [x] Only shows Program NCAs in Application list
- [x] Error popup for certain errors
- [x] Change shown GUI columns without restarting
- [x] Sort by column if you click on the column header
  - [x] Fix sort on `Time Played`, `Last Played` and `File Size` columns
- [x] Add ability to designate favourite games
- [x] Load metadata from JSONs
- [x] Implement async loading of the game list
- [x] Progress bar and information about what is being loaded

Xpl0itR commented:
I guess that's everything, ready for review.
Also temp controller images need to be replaced before merging.

emmauss commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -423,10 +457,40 @@ private static IAalOutput InitializeAudioEngine()
+
+            using (Stream stream = File.OpenRead(metadataPath))
+            {
+                AppMetadata = JsonSerializer.Deserialize<ApplicationMetadata>(stream, resolver);
```
this should be a local variable, as it's data is only used within this scope.

emmauss commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -534,67 +600,179 @@ private void About_Pressed(object o, EventArgs args)
+            string aValue = model.GetValue(a, 5).ToString();
+            string bValue = model.GetValue(b, 5).ToString();
+
+            if (aValue.Length > 4 && aValue.Substring(aValue.Length - 4) == "mins")
```
you could easily do the sorting if TimePlayed is a TimeSpan object.

Xpl0itR replied:
The problem is that the data input into the sorting function is the string shown in the gui

manotroll commented:
where is the localization files the portuguese is not working

Thog commented on Ryujinx/_schema.json:
```diff
@@ -496,17 +496,17 @@
       "title": "Used to toggle columns in the GUI",
       "description": "Used to toggle columns in the GUI",
-      "default": [ true, true, true, true, true, true, true, true, true ]
+      "default": [ true, true, true, true, true, true, true, true, true, true ]
```
This doesn't look right, this should be an array of object like here https://github.com/Ryujinx/Ryujinx/pull/757/files#diff-a40c86305da6aa451df7a23c3465b704R24

Thog commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -457,4 +473,11 @@ private static byte[] NspOrXciIcon(string applicationPath)
+    {
+        public ApplicationLibrary.ApplicationData AppData { get; set; }
+        public float NumAppsFound  { get; set; }
+        public float NumAppsLoaded { get; set; }
```
Align thoses

Thealexbarney commented:
Changes look good to me. That leaves the freezing or hanging when sorting or reloading games

AcK77 commented on Ryujinx/Config.json:
```diff
@@ -72,6 +89,7 @@
         },
         "right_joycon": {
             "stick": "Axis3",
+            "stick_y": "Button0",
```
Is it really a thing ?

Xpl0itR replied:
as far as I know its unused. I didn't manually add that. I will remove

AcK77 commented on Ryujinx/Ryujinx.csproj:
```diff
@@ -23,8 +23,35 @@
+    <None Remove="Ui\assets\BlueCon.png" />
+    <None Remove="Ui\assets\ProCon.png" />
+    <None Remove="Ui\assets\RedCon.png" />
+    <None Remove="Ui\assets\ryujinxNCAIcon.png" />
```
Why some file names start with `ryujinx`? As example maybe `NCAIcon` is fine and every file names start with an uppercase letter.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -1,66 +1,73 @@
 
-        public static List<ApplicationData> ApplicationLibraryData { get; private set; }
+        private static Keyset _keySet;
+        private static SystemState.TitleLanguage _desiredTitleLanguage;
```
```csharp
        private static Keyset                    _keySet;
        private static SystemState.TitleLanguage _desiredTitleLanguage;
```

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -1,66 +1,73 @@
-        private const double SecondsPerMinute = 60.0;
-        private const double SecondsPerHour   = SecondsPerMinute * 60;
-        private const double SecondsPerDay    = SecondsPerHour   * 24;
+        public struct ApplicationData
```
Maybe you should store struct in a different file to make the code more readable.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -1,66 +1,73 @@
+        private static SystemState.TitleLanguage _desiredTitleLanguage;
 
-        public struct ApplicationData
+        private struct ApplicationMetadata
```
Maybe you should store struct in a different file to make the code more readable.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -1,66 +1,73 @@
         }
 
-        public static void Init(List<string> AppDirs, Keyset keySet, SystemState.TitleLanguage desiredTitleLanguage)
+        private static ApplicationMetadata _appMetadata;
```
(Don't forget to align the var with both above if you remove the struct from this file)

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -1,66 +1,73 @@
-            RyujinxNroIcon = GetResourceBytes("Ryujinx.Ui.assets.ryujinxNROIcon.png");
-            RyujinxNsoIcon = GetResourceBytes("Ryujinx.Ui.assets.ryujinxNSOIcon.png");
+            _keySet = keySet;
+            _desiredTitleLanguage = desiredTitleLanguage;
```
```csharp
            _keySet               = keySet;
            _desiredTitleLanguage = desiredTitleLanguage;
```

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -69,30 +76,77 @@ public static void Init(List<string> AppDirs, Keyset keySet, SystemState.TitleLa
-                DirectoryInfo AppDirInfo = new DirectoryInfo(appDir);
-                foreach (FileInfo App in AppDirInfo.GetFiles())
+                string[] apps = Directory.GetFiles(appDir, "*.*", SearchOption.AllDirectories);
+                foreach (string app in apps)
```
```suggestion
                foreach (string app in Directory.GetFiles(appDir, "*.*", SearchOption.AllDirectories))
```
and remove `string[] apps` since you don't use it at another place.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -69,30 +76,77 @@ public static void Init(List<string> AppDirs, Keyset keySet, SystemState.TitleLa
+                            PartitionFileSystem nsp = new PartitionFileSystem(new FileStream(app, FileMode.Open, FileAccess.Read).AsStorage());
+                            foreach (DirectoryEntryEx fileEntry in nsp.EnumerateEntries("/", "*.nca"))
+                            {
+                                nsp.OpenFile(out IFile ncaFile, fileEntry.FullPath, OpenMode.Read).ThrowIfFailure();
```
Add a new empty line after.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -69,30 +76,77 @@ public static void Init(List<string> AppDirs, Keyset keySet, SystemState.TitleLa
+                            {
+                                nsp.OpenFile(out IFile ncaFile, fileEntry.FullPath, OpenMode.Read).ThrowIfFailure();
+                                Nca nca = new Nca(_keySet, ncaFile.AsStorage());
+                                int dataIndex = Nca.GetSectionIndexFromType(NcaSectionType.Data, NcaContentType.Program);
```
```csharp
                                Nca nca       = new Nca(_keySet, ncaFile.AsStorage());
                                int dataIndex = Nca.GetSectionIndexFromType(NcaSectionType.Data, NcaContentType.Program);
```

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -69,30 +76,77 @@ public static void Init(List<string> AppDirs, Keyset keySet, SystemState.TitleLa
+                                }
+                            }
+
+                            if (!hasMainNca) continue;
```
```suggestion
                            if (!hasMainNca) 
                            {
                                continue;
                            }
```

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -69,30 +76,77 @@ public static void Init(List<string> AppDirs, Keyset keySet, SystemState.TitleLa
+                        try
+                        {
+                            Nca nca = new Nca(_keySet, new FileStream(app, FileMode.Open, FileAccess.Read).AsStorage());
+                            int dataIndex = Nca.GetSectionIndexFromType(NcaSectionType.Data, NcaContentType.Program);
```
Align assignments as my previous comment.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -69,30 +76,77 @@ public static void Init(List<string> AppDirs, Keyset keySet, SystemState.TitleLa
+                string titleName       = "Unknown";
+                string titleId         = "0000000000000000";
+                string developer       = "Unknown";
+                string version         = "?";
```
Maybe the version should be `0` by default ?

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -103,239 +157,234 @@ public static void Init(List<string> AppDirs, Keyset keySet, SystemState.TitleLa
-                            {
-                                // Creates NACP class from the memory stream
-                                Nacp controlData = new Nacp(stream);
+                            using MemoryStream stream = new MemoryStream(Read(assetOffset + (int)nacpOffset, (int)nacpSize));
```
Why you removed the `()` and the brackets of the using?

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -103,239 +157,234 @@ public static void Init(List<string> AppDirs, Keyset keySet, SystemState.TitleLa
                 }
 
-                string[] playedData = GetPlayedData(titleId, "00000000000000000000000000000001");
+                (bool fav, string timePlayed, string lastPlayed) = GetMetadata(titleId);
```
```suggestion
                (bool favorite, string timePlayed, string lastPlayed) = GetMetadata(titleId);
```
for consistency

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -377,84 +426,71 @@ private static IFileSystem GetControlFs(PartitionFileSystem Pfs)
         }
 
-        private static string[] GetPlayedData(string TitleId, string UserId)
+        private static (bool fav, string timePlayed, string lastPlayed) GetMetadata(string titleId)
```
```suggestion
        private static (bool favorite, string timePlayed, string lastPlayed) GetMetadata(string titleId)
```

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -377,84 +426,71 @@ private static IFileSystem GetControlFs(PartitionFileSystem Pfs)
-            if (Path.GetExtension(applicationPath) == ".xci")
+            const int secondsPerMinute = 60;
+            const int secondsPerHour   = secondsPerMinute * 60;
+            const int secondsPerDay    = secondsPerHour * 24;
```
```suggestion
            const int secondsPerDay    = secondsPerHour   * 24;
```

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -377,84 +426,71 @@ private static IFileSystem GetControlFs(PartitionFileSystem Pfs)
+            const int secondsPerMinute = 60;
+            const int secondsPerHour   = secondsPerMinute * 60;
+            const int secondsPerDay    = secondsPerHour * 24;
+            string readableString;
```
Add a new line before.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -377,84 +426,71 @@ private static IFileSystem GetControlFs(PartitionFileSystem Pfs)
         }
     }
+
+    public class ApplicationAddedEventArgs : EventArgs
```
Maybe you should store class in a different file to make the code more readable.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -377,84 +426,71 @@ private static IFileSystem GetControlFs(PartitionFileSystem Pfs)
+    {
+        public ApplicationData AppData { get; set; }
+        public int NumAppsFound        { get; set; }
+        public int NumAppsLoaded       { get; set; }
```
Align names.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -13,66 +15,109 @@
 
 namespace Ryujinx.UI
 {
+    public struct GuiColumns
```
Maybe you should store struct in a different file to make the code more readable.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -13,66 +15,109 @@
-        private static string _userId = "00000000000000000000000000000001";
+        private static TreeView _treeView;
+
+        private struct ApplicationMetadata
```
Maybe you should store struct in a different file to make the code more readable.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -94,110 +139,123 @@ private MainWindow(Builder builder, string[] args, Application gtkApplication) :
         {
-            _tableStore.Clear();
-            ApplicationLibrary.Init(SwitchSettings.SwitchConfig.GameDirs, _device.System.KeySet, _device.System.State.DesiredTitleLanguage);
+            if (!SwitchSettings.SwitchConfig.EnableCustomTheme) return;
```
```suggestion
            if (!SwitchSettings.SwitchConfig.EnableCustomTheme)
            {
                return;
            }
```

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -94,110 +139,123 @@ private MainWindow(Builder builder, string[] args, Application gtkApplication) :
+            foreach (TreeViewColumn column in _gameTable.Columns)
             {
-                cssProvider.LoadFromPath(System.IO.Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Theme.css"));
+                     if (column.Title == "Fav")         { _favColumn        = column; }
```
```csharp
                if (column.Title == "Fav")              { _favColumn        = column; }
                else if (column.Title == "Application") { _appColumn        = column; }
                else if (column.Title == "Developer")   { _devColumn        = column; }
                ...
```

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -94,110 +139,123 @@ private MainWindow(Builder builder, string[] args, Application gtkApplication) :
+
+        internal static async Task UpdateGameTable()
+        {
+            if (_updatingGameTable) return;
```
```suggestion
            if (_updatingGameTable)
            {
                return;
            }
```

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -94,110 +139,123 @@ private MainWindow(Builder builder, string[] args, Application gtkApplication) :
+
+            _updatingGameTable         = true;
+            _treeView.HeadersClickable = false;
+            _tableStore.Clear();
```
Add a new line before.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -306,92 +368,89 @@ internal void LoadApplication(string path)
-                        }
-                    }
+                IJsonFormatterResolver resolver = CompositeResolver.Create(new[] { StandardResolver.AllowPrivateSnakeCase });
+                ApplicationMetadata appMetadata;
```
Add a new line before.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -306,92 +368,89 @@ internal void LoadApplication(string path)
 
         private static void End()
         {
+            if (_ending) return;
```
```suggestion
            if (_ending)
            {
                return;
            }
```

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -306,92 +368,89 @@ internal void LoadApplication(string path)
+                string metadataFile   = System.IO.Path.Combine(metadataFolder, "metadata.json");
+
+                IJsonFormatterResolver resolver = CompositeResolver.Create(new[] { StandardResolver.AllowPrivateSnakeCase });
+                ApplicationMetadata appMetadata;
```
Add a new line before.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -423,15 +482,56 @@ private static IAalOutput InitializeAudioEngine()
+
+        private void FavToggle_Toggled(object sender, ToggledArgs args)
+        {
+            _tableStore.GetIter(out TreeIter treeIter, new TreePath(args.Path));
```
Add a new line after.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -423,15 +482,56 @@ private static IAalOutput InitializeAudioEngine()
+            string metadataPath = System.IO.Path.Combine(new VirtualFileSystem().GetBasePath(), "games", titleid, "gui", "metadata.json");
+
+            IJsonFormatterResolver resolver = CompositeResolver.Create(new[] { StandardResolver.AllowPrivateSnakeCase });
+            ApplicationMetadata appMetadata;
```
Add a new line before.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -524,77 +624,204 @@ private void Update_Pressed(object o, EventArgs args)
-            _gtkApplication.AddWindow(AboutWin);
+        private void Fav_Toggled(object sender, EventArgs args)
+        {
+            GuiColumns updatedColumns = SwitchSettings.SwitchConfig.GuiColumns;
```
Add a new line after. (same for all other `Toggled` functions)

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -524,77 +624,204 @@ private void Update_Pressed(object o, EventArgs args)
+        {
+            GuiColumns updatedColumns = SwitchSettings.SwitchConfig.GuiColumns;
+            updatedColumns.FavColumn  = _favToggle.Active;
+            SwitchSettings.SwitchConfig.GuiColumns = updatedColumns;
```
Align assigns. (same for all other `Toggled` functions)

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -524,77 +624,204 @@ private void Update_Pressed(object o, EventArgs args)
+                bValue = bValue.Substring(0, bValue.Length - 1);
+            }
+
+            if (float.Parse(aValue) > float.Parse(bValue)) return -1;
```
Add and expand brackets on all those conditions.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -524,77 +624,204 @@ private void Update_Pressed(object o, EventArgs args)
+            string aValue = model.GetValue(a, 6).ToString();
+            string bValue = model.GetValue(b, 6).ToString();
+
+            if (aValue == "Never") aValue = DateTime.UnixEpoch.ToString();
```
Add and expand bracket. (Same for the line just above)

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -524,77 +624,204 @@ private void Update_Pressed(object o, EventArgs args)
+            {
+                aValue = (float.Parse(aValue[0..^2]) * 1024).ToString();
+            }
+            else aValue = aValue[0..^2];
```
Add and expand bracket.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -524,77 +624,204 @@ private void Update_Pressed(object o, EventArgs args)
+            {
+                bValue = (float.Parse(bValue[0..^2]) * 1024).ToString();
+            }
+            else bValue = bValue[0..^2];
```
Add and expand bracket.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -524,77 +624,204 @@ private void Update_Pressed(object o, EventArgs args)
+            }
+            else bValue = bValue[0..^2];
+
+                 if (float.Parse(aValue) > float.Parse(bValue)) return -1;
```
Add and expand bracket on all those conditions.

Xpl0itR commented:
Done

AcK77 commented on Ryujinx/Ui/AboutWindow.cs:
```diff
@@ -10,35 +10,37 @@
     public class AboutWindow : Window
     {
-        public static Info Information { get; private set; }
+        private struct Info
```
You should store struct in another file for consistency.

Thealexbarney replied:
I don't think you can put a nested private struct in another file

Xpl0itR commented:
done

gdkchan commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -103,158 +137,164 @@ public static void Init(List<string> AppDirs, Keyset keySet, SystemState.TitleLa
 
                             // Creates memory stream out of byte array which is the NACP
-                            using (MemoryStream stream = new MemoryStream(Read(AssetOffset + (int)nacpOffset, (int)nacpSize)))
+                            using (MemoryStream stream = new MemoryStream(Read(assetOffset + (int) nacpOffset, (int) nacpSize)))
```
`(int) ` Why the space?

gdkchan commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -377,84 +408,65 @@ private static IFileSystem GetControlFs(PartitionFileSystem Pfs)
         }
 
-        private static string[] GetPlayedData(string TitleId, string UserId)
+        private static (bool favorite, string timePlayed, string lastPlayed) GetMetadata(string titleId)
```
I recommend using a struct here instead of a tuple.

Xpl0itR replied:
I disagree, imo that would just add complication with no visible benefit

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -94,110 +119,128 @@ private MainWindow(Builder builder, string[] args, Application gtkApplication) :
+            if (SwitchSettings.SwitchConfig.GuiColumns.FileSizeColumn)   { _fileSizeToggle.Active   = true; }
+            if (SwitchSettings.SwitchConfig.GuiColumns.PathColumn)       { _pathToggle.Active       = true; }
+
+            _gameTable.Model = _tableStore = new ListStore(typeof(bool), typeof(Gdk.Pixbuf), typeof(string), typeof(string), typeof(string), typeof(string), typeof(string), typeof(string), typeof(string), typeof(string));
```
I recommend adding line breaks so that each typeof is on a new line.

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -94,110 +119,128 @@ private MainWindow(Builder builder, string[] args, Application gtkApplication) :
                 Title          = "Ryujinx - Error",
-                Icon           = new Gdk.Pixbuf(Assembly.GetExecutingAssembly(), "Ryujinx.Ui.assets.ryujinxIcon.png"),
+                Icon           = new Gdk.Pixbuf(Assembly.GetExecutingAssembly(), "Ryujinx.Ui.assets.Icon.png"),
+                Text           = "Ryujinx has encountered an Error",
```
Why is Error upper cased? Shouldn't it end with a dot?

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -94,110 +119,128 @@ private MainWindow(Builder builder, string[] args, Application gtkApplication) :
+            else
             {
-                _tableStore.AppendValues(new Gdk.Pixbuf(AppData.Icon, 75, 75), $"{AppData.TitleName}\n{AppData.TitleId.ToUpper()}", AppData.Developer, AppData.Version, AppData.TimePlayed, AppData.LastPlayed, AppData.FileExt, AppData.FileSize, AppData.Path);
+                Logger.PrintWarning(LogClass.Application, $"The \"custom_theme_path\" section in \"Config.json\" contains an invalid path: \"{SwitchSettings.SwitchConfig.CustomThemePath}\"");
```
Messages should end with a dot.

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -94,110 +119,128 @@ private MainWindow(Builder builder, string[] args, Application gtkApplication) :
         }
 
-        public static void CreateErrorDialog(string errorMessage)
+        internal static void CreateErrorDialog(string errorMessage)
```
Why is it internal instead of private? Is anything else using it outside of MainWindow?

Xpl0itR replied:
yes its used in program.cs

gdkchan replied:
I'd recommend moving to a separate helper class then.

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -306,92 +353,94 @@ internal void LoadApplication(string path)
+                    appMetadata = JsonSerializer.Deserialize<ApplicationMetadata>(stream, resolver);
                 }
+
+                appMetadata.TimePlayed += Math.Round(DateTime.UtcNow.Subtract(DateTime.Parse(appMetadata.LastPlayed)).TotalSeconds, MidpointRounding.AwayFromZero);
```
Break it down into multiple lines.

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -423,15 +472,58 @@ private static IAalOutput InitializeAudioEngine()
+        private void Application_Added(object sender, ApplicationAddedEventArgs e)
+        {
+            Application.Invoke(delegate {
+                _tableStore.AppendValues(e.AppData.Favorite, new Gdk.Pixbuf(e.AppData.Icon, 75, 75), $"{e.AppData.TitleName}\n{e.AppData.TitleId.ToUpper()}", e.AppData.Developer, e.AppData.Version, e.AppData.TimePlayed, e.AppData.LastPlayed, e.AppData.FileExtension, e.AppData.FileSize, e.AppData.Path);
```
Break it down into multiple lines.

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -423,15 +472,58 @@ private static IAalOutput InitializeAudioEngine()
-        private void Row_Activated(object o, RowActivatedArgs args)
+        private void Application_Added(object sender, ApplicationAddedEventArgs e)
+        {
+            Application.Invoke(delegate {
```
`{` should be on a new line.

Xpl0itR replied:
smh Thog, this is c# not c++ :p

Thog replied:
~~just my Java codestyle coming back~~

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -524,77 +616,245 @@ private void Update_Pressed(object o, EventArgs args)
+#pragma warning restore CS4014
+        }
+
+        private static int TimePlayedSort(ITreeModel model, TreeIter a, TreeIter b)
```
That methods seems to be really fragile to be honest. If someone inadvertently changes "hrs" to "hours" for example, the code breaks. Any slightly change on the string would break the sort. Surely there must be a better way to do that...

Xpl0itR replied:
The treeview sorts the string present in the field, afaik there isn't a better way to do this

Xpl0itR commented:
should be all good now
- completed gdkchan's requested changes that have been marked resolved
- made some structs internal as they aren't used outside of the GUI
- renamed Ryujinx.UI to Ryujinx.Ui to fit naming convention and folder structure
- fixed bug where controller type dropdown box is stretched
